File: ios/Classes/TelemetryAggregator.swift
import Foundation
import UIKit

/// Aggregates all session telemetry in the native layer (D016).
/// Each platform bridge (Flutter, RN, future) only calls getSessionTelemetry()
/// and POSTs the result — no per-bridge telemetry reimplementation needed.
///
/// Thread-safe: all mutable state accessed via syncQueue.
class TelemetryAggregator {

    private let syncQueue = DispatchQueue(label: "io.polyfence.telemetryAggregator", attributes: .concurrent)

    // --- Activity distribution ---
    private var activityTimeMs: [String: Int64] = [:]
    private var lastActivityChangeTime: Int64
    private var lastTrackedActivity: String = "unknown"

    // --- GPS interval distribution ---
    private var intervalTimeMs: [Int64: Int64] = [:]
    private var lastIntervalChangeTime: Int64
    private var lastTrackedInterval: Int64 = 10_000
    private var totalIntervalMs: Int64 = 0
    private var intervalSampleCount: Int = 0

    // --- Stationary tracking ---
    private var cumulativeStationaryMs: Int64 = 0
    private var stationaryStartTime: Int64?

    // --- GPS accuracy at events ---
    private var eventAccuracies: [Double] = []
    private var eventSpeeds: [Double] = []
    private var boundaryEventsCount: Int = 0
    private let boundaryThresholdM: Double = 50.0

    // --- False event tracking ---
    private var lastEventPerZone: [String: (String, Int64)] = [:]
    private var falseEventCount: Int = 0

    // --- Detection timing ---
    private var detectionTimesMs: [Double] = []
    private var firstDetectionTimeMs: Int64?
    private var sessionStartTime: Int64

    // --- Zone metrics ---
    private var zoneTransitionCount: Int = 0
    private var dwellDurations: [Double] = []

    // --- GPS health ---
    private var totalGpsReadings: Int = 0
    private var goodGpsReadings: Int = 0

    // --- Error tracking ---
    private var errorCounts: [String: Int] = [:]

    // --- Device & config info ---
    private var deviceCategory: String?
    private var osVersionMajor: Int
    private var batteryLevelStart: Double?
    private var batteryLevelEnd: Double?
    private var chargingDuringSession: Bool = false
    private var accuracyProfile: String?
    private var updateStrategy: String?

    init() {
        let now = Int64(Date().timeIntervalSince1970 * 1000)
        lastActivityChangeTime = now
        lastIntervalChangeTime = now
        sessionStartTime = now
        osVersionMajor = ProcessInfo.processInfo.operatingSystemVersion.majorVersion
    }

    private func currentTimeMs() -> Int64 {
        return Int64(Date().timeIntervalSince1970 * 1000)
    }

    // ========================================================================
    // RECORDING METHODS — called by engines during session
    // ========================================================================

    /// Record an activity change. Accumulates time spent in each activity type.
    func recordActivityChange(activityType: String) {
        syncQueue.async(flags: .barrier) {
            let now = self.currentTimeMs()
            let elapsed = now - self.lastActivityChangeTime
            if elapsed > 0 {
                self.activityTimeMs[self.lastTrackedActivity, default: 0] += elapsed
            }
            self.lastActivityChangeTime = now
            self.lastTrackedActivity = activityType.lowercased()
        }
    }

    /// Finalize the last activity segment. Must be called before reading distribution.
    func finalizeActivityTracking() {
        recordActivityChange(activityType: lastTrackedActivity)
    }

    /// Record a GPS update with the current interval and accuracy.
    func recordGpsUpdate(intervalMs: Int64, accuracyM: Float) {
        syncQueue.async(flags: .barrier) {
            let now = self.currentTimeMs()
            let elapsed = now - self.lastIntervalChangeTime
            if elapsed > 0 {
                self.intervalTimeMs[self.lastTrackedInterval, default: 0] += elapsed
            }
            self.lastIntervalChangeTime = now
            self.lastTrackedInterval = intervalMs
            self.totalIntervalMs += intervalMs
            self.intervalSampleCount += 1

            self.totalGpsReadings += 1
            if accuracyM <= 100.0 {
                self.goodGpsReadings += 1
            }
        }
    }

    /// Record a geofence event with context.
    func recordGeofenceEvent(
        zoneId: String,
        eventType: String,
        distanceM: Double,
        speedMps: Double,
        accuracyM: Double,
        detectionTimeMs: Double
    ) {
        syncQueue.async(flags: .barrier) {
            self.zoneTransitionCount += 1
            self.detectionTimesMs.append(detectionTimeMs)

            if self.firstDetectionTimeMs == nil {
                self.firstDetectionTimeMs = self.currentTimeMs() - self.sessionStartTime
            }

            self.eventAccuracies.append(accuracyM)
            self.eventSpeeds.append(speedMps)

            if distanceM >= 0 && distanceM < self.boundaryThresholdM {
                self.boundaryEventsCount += 1
            }

            // False event detection (reversal within 30s on same zone)
            if let lastEvent = self.lastEventPerZone[zoneId] {
                let (lastType, lastTime) = lastEvent
                let timeSince = self.currentTimeMs() - lastTime
                if timeSince <= 30_000 && lastType != eventType {
                    self.falseEventCount += 1
                }
            }
            self.lastEventPerZone[zoneId] = (eventType, self.currentTimeMs())
        }
    }

    /// Record dwell completion (on EXIT).
    func recordDwellComplete(durationMinutes: Double) {
        syncQueue.async(flags: .barrier) {
            self.dwellDurations.append(durationMinutes)
        }
    }

    /// Record stationary state change.
    func recordStationaryChange(isStationary: Bool) {
        syncQueue.async(flags: .barrier) {
            if isStationary && self.stationaryStartTime == nil {
                self.stationaryStartTime = self.currentTimeMs()
            } else if !isStationary, let start = self.stationaryStartTime {
                self.cumulativeStationaryMs += self.currentTimeMs() - start
                self.stationaryStartTime = nil
            }
        }
    }

    /// Record an error occurrence.
    func recordError(errorType: String) {
        syncQueue.async(flags: .barrier) {
            self.errorCounts[errorType, default: 0] += 1
        }
    }

    /// Set device info (call once at session start).
    func setDeviceInfo(category: String, osVersion: Int) {
        syncQueue.async(flags: .barrier) {
            self.deviceCategory = category
            self.osVersionMajor = osVersion
        }
    }

    /// Set battery info.
    func setBatteryInfo(startPercent: Double?, endPercent: Double?, chargingDuring: Bool) {
        syncQueue.async(flags: .barrier) {
            if let s = startPercent { self.batteryLevelStart = s }
            if let e = endPercent { self.batteryLevelEnd = e }
            self.chargingDuringSession = chargingDuring
        }
    }

    /// Set configuration context.
    func setConfig(accuracyProfile: String, updateStrategy: String) {
        syncQueue.async(flags: .barrier) {
            self.accuracyProfile = accuracyProfile
            self.updateStrategy = updateStrategy
        }
    }

    // ========================================================================
    // OUTPUT — returns complete v2 enhanced payload
    // ========================================================================

    /// Returns the complete v2 enhanced session telemetry payload.
    /// Matches the schema defined in intelligence/DATA_STRATEGY.md.
    func getSessionTelemetry(
        geofenceEngine: GeofenceEngine? = nil
    ) -> [String: Any] {
        var result: [String: Any] = [:]

        syncQueue.sync {
            // Finalize activity tracking
            let now = currentTimeMs()
            let actElapsed = now - lastActivityChangeTime
            var activitySnapshot = activityTimeMs
            if actElapsed > 0 {
                activitySnapshot[lastTrackedActivity, default: 0] += actElapsed
            }

            // Finalize stationary
            var totalStationary = cumulativeStationaryMs
            if let start = stationaryStartTime {
                totalStationary += now - start
            }

            let sessionDurationMs = now - sessionStartTime
            let sessionDurationMinutes = Double(sessionDurationMs) / 60_000.0

            // Activity distribution
            let totalActivityMs = activitySnapshot.values.reduce(0, +)
            var activityDist: [String: Double] = [:]
            if totalActivityMs > 0 {
                for (key, ms) in activitySnapshot {
                    activityDist[key] = Double(ms) / Double(totalActivityMs)
                }
            }

            // GPS interval distribution
            var intervalSnapshot = intervalTimeMs
            let intElapsed = now - lastIntervalChangeTime
            if intElapsed > 0 {
                intervalSnapshot[lastTrackedInterval, default: 0] += intElapsed
            }
            let totalIntMs = intervalSnapshot.values.reduce(0, +)
            var intervalDist: [String: Double] = [:]
            if totalIntMs > 0 {
                for (key, ms) in intervalSnapshot {
                    intervalDist[String(key)] = Double(ms) / Double(totalIntMs)
                }
            }

            // Detection stats
            let detTimes = detectionTimesMs
            let detAvg = detTimes.isEmpty ? 0.0 : detTimes.reduce(0, +) / Double(detTimes.count)
            let detP95: Double
            if detTimes.isEmpty {
                detP95 = 0.0
            } else {
                let sorted = detTimes.sorted()
                let idx = min(Int(Double(sorted.count) * 0.95), sorted.count - 1)
                detP95 = sorted[idx]
            }

            // Event averages
            let avgAccuracy = eventAccuracies.isEmpty ? 0.0 : eventAccuracies.reduce(0, +) / Double(eventAccuracies.count)
            let avgSpeed = eventSpeeds.isEmpty ? 0.0 : eventSpeeds.reduce(0, +) / Double(eventSpeeds.count)

            // False event ratio
            let totalDetections = detTimes.count
            let falseRatio = totalDetections > 0 ? Double(falseEventCount) / Double(totalDetections) : 0.0

            // Dwell average
            let avgDwell = dwellDurations.isEmpty ? 0.0 : dwellDurations.reduce(0, +) / Double(dwellDurations.count)

            // Stationary ratio
            let statRatio = sessionDurationMs > 0 ? Double(totalStationary) / Double(sessionDurationMs) : 0.0

            // Average GPS interval
            let avgInterval = intervalSampleCount > 0 ? Int(totalIntervalMs / Int64(intervalSampleCount)) : 0

            // GPS OK ratio
            let gpsOk = totalGpsReadings > 0 ? Double(goodGpsReadings) / Double(totalGpsReadings) : 0.0

            var telemetry = SessionTelemetry()
            telemetry.detectionsTotal = totalDetections
            telemetry.detectionTimeAvgMs = detAvg
            telemetry.detectionTimeP95Ms = detP95
            telemetry.gpsAccuracyAvgM = avgAccuracy
            telemetry.sessionDurationMinutes = sessionDurationMinutes
            telemetry.errorCounts = errorCounts
            telemetry.ttfdMs = firstDetectionTimeMs ?? 0
            telemetry.hadDetection = totalDetections > 0
            telemetry.gpsOkRatio = gpsOk
            telemetry.sampleEvents = totalGpsReadings
            telemetry.batteryLevelStart = batteryLevelStart
            telemetry.batteryLevelEnd = batteryLevelEnd
            telemetry.accuracyProfile = accuracyProfile
            telemetry.updateStrategy = updateStrategy
            telemetry.activityDistribution = activityDist
            telemetry.gpsIntervalDistribution = intervalDist
            telemetry.stationaryRatio = statRatio
            telemetry.avgGpsIntervalMs = avgInterval
            telemetry.falseEventCount = falseEventCount
            telemetry.falseEventRatio = falseRatio
            telemetry.avgGpsAccuracyAtEvent = avgAccuracy
            telemetry.avgSpeedAtEventMps = avgSpeed
            telemetry.boundaryEventsCount = boundaryEventsCount
            telemetry.zoneTransitionCount = zoneTransitionCount
            telemetry.avgDwellMinutes = avgDwell
            telemetry.deviceCategory = deviceCategory
            telemetry.osVersionMajor = osVersionMajor
            telemetry.chargingDuringSession = chargingDuringSession

            result = telemetry.toMap()
        }

        // Zone metrics from engine (outside sync to avoid deadlock)
        if let engine = geofenceEngine {
            result["zone_count"] = engine.getZoneCount()
            result["zone_size_distribution"] = engine.getZoneSizeDistribution()
        }

        return result
    }

    /// Reset all telemetry for a new session.
    func resetTelemetry() {
        syncQueue.async(flags: .barrier) {
            let now = self.currentTimeMs()

            self.activityTimeMs.removeAll()
            self.lastActivityChangeTime = now
            self.lastTrackedActivity = "unknown"

            self.intervalTimeMs.removeAll()
            self.lastIntervalChangeTime = now
            self.lastTrackedInterval = 10_000
            self.totalIntervalMs = 0
            self.intervalSampleCount = 0

            self.cumulativeStationaryMs = 0
            self.stationaryStartTime = nil

            self.eventAccuracies.removeAll()
            self.eventSpeeds.removeAll()
            self.boundaryEventsCount = 0

            self.lastEventPerZone.removeAll()
            self.falseEventCount = 0

            self.detectionTimesMs.removeAll()
            self.firstDetectionTimeMs = nil
            self.sessionStartTime = now

            self.zoneTransitionCount = 0
            self.dwellDurations.removeAll()

            self.totalGpsReadings = 0
            self.goodGpsReadings = 0

            self.errorCounts.removeAll()

            self.batteryLevelStart = nil
            self.batteryLevelEnd = nil
            self.chargingDuringSession = false
        }
    }

    // ========================================================================
    // DEVICE CATEGORY HELPERS
    // ========================================================================

    /// Determine device category. Returns a bucketed category — NOT the exact model.
    static func getDeviceCategory() -> String {
        var systemInfo = utsname()
        uname(&systemInfo)
        let machine = withUnsafePointer(to: &systemInfo.machine) {
            $0.withMemoryRebound(to: CChar.self, capacity: 1) {
                String(validatingUTF8: $0) ?? "unknown"
            }
        }

        if machine.hasPrefix("iPhone") {
            let stripped = machine.replacingOccurrences(of: "iPhone", with: "")
            let parts = stripped.split(separator: ",")
            if let major = Int(parts.first ?? "") {
                if major >= 15 { return "iphone_flagship" }
                if major >= 12 { return "iphone_standard" }
                return "iphone_older"
            }
        }
        if machine.hasPrefix("iPad") { return "ipad" }
        return "ios_other"
    }
}

File: android/src/main/kotlin/io/polyfence/core/TelemetryAggregator.kt
package io.polyfence.core

import android.os.Build
import java.util.concurrent.ConcurrentHashMap

/**
 * Aggregates all session telemetry in the native layer (D016).
 * Each platform bridge (Flutter, RN, future) only calls getSessionTelemetry()
 * and POSTs the result — no per-bridge telemetry reimplementation needed.
 *
 * Thread-safe: all mutable state is accessed under a single lock object,
 * matching the Swift implementation's DispatchQueue(attributes: .concurrent)
 * with barrier writes and sync reads.
 */
internal class TelemetryAggregator {

    private val lock = Any()

    // --- Activity distribution ---
    private val activityTimeMs = mutableMapOf<String, Long>()
    private var lastActivityChangeTime: Long = System.currentTimeMillis()
    private var lastTrackedActivity: String = "unknown"

    // --- GPS interval distribution ---
    private val intervalTimeMs = mutableMapOf<Long, Long>()
    private var lastIntervalChangeTime: Long = System.currentTimeMillis()
    private var lastTrackedInterval: Long = 10_000L
    private var totalIntervalMs: Long = 0L
    private var intervalSampleCount: Int = 0

    // --- Stationary tracking ---
    private var cumulativeStationaryMs: Long = 0L
    private var stationaryStartTime: Long? = null

    // --- GPS accuracy at events ---
    private val eventAccuracies = mutableListOf<Double>()
    private val eventSpeeds = mutableListOf<Double>()
    private val eventDistances = mutableListOf<Double>()
    private var boundaryEventsCount: Int = 0
    private val BOUNDARY_THRESHOLD_M = 50.0

    // --- False event tracking ---
    private val lastEventPerZone = ConcurrentHashMap<String, Pair<String, Long>>()
    private var falseEventCount: Int = 0

    // --- Detection timing ---
    private val detectionTimesMs = mutableListOf<Double>()
    private var firstDetectionTimeMs: Long? = null
    private var sessionStartTime: Long = System.currentTimeMillis()
    private var sessionStartHour: Int = java.util.Calendar.getInstance().get(java.util.Calendar.HOUR_OF_DAY)

    // --- Zone metrics ---
    private var zoneTransitionCount: Int = 0
    private val dwellDurations = mutableListOf<Double>()

    // --- GPS health ---
    private var totalGpsReadings: Int = 0
    private var goodGpsReadings: Int = 0

    // --- Error tracking ---
    private val errorCounts = mutableMapOf<String, Int>()

    // --- Device & config info ---
    private var deviceCategory: String? = null
    private var osVersionMajor: Int = Build.VERSION.SDK_INT
    private var batteryLevelStart: Double? = null
    private var batteryLevelEnd: Double? = null
    private var chargingDuringSession: Boolean = false
    private var accuracyProfile: String? = null
    private var updateStrategy: String? = null

    // ========================================================================
    // RECORDING METHODS — called by engines during session
    // ========================================================================

    /**
     * Record an activity change. Accumulates time spent in each activity type.
     * Call BEFORE updating the current activity.
     */
    fun recordActivityChange(activityType: String) {
        synchronized(lock) {
            val now = System.currentTimeMillis()
            val elapsed = now - lastActivityChangeTime
            if (elapsed > 0) {
                activityTimeMs[lastTrackedActivity] =
                    (activityTimeMs[lastTrackedActivity] ?: 0L) + elapsed
            }
            lastActivityChangeTime = now
            lastTrackedActivity = activityType.lowercase()
        }
    }

    /**
     * Finalize the last activity segment. Useful for callers who want to
     * explicitly close the current activity window. Note: getSessionTelemetry()
     * snapshots activity data without calling this, so it is safe to call
     * getSessionTelemetry() without finalizing first.
     */
    fun finalizeActivityTracking() {
        recordActivityChange(lastTrackedActivity)
    }

    /**
     * Record a GPS update with the current interval and accuracy.
     */
    fun recordGpsUpdate(intervalMs: Long, accuracyM: Float) {
        synchronized(lock) {
            // Interval histogram
            val now = System.currentTimeMillis()
            val elapsed = now - lastIntervalChangeTime
            if (elapsed > 0) {
                intervalTimeMs[lastTrackedInterval] =
                    (intervalTimeMs[lastTrackedInterval] ?: 0L) + elapsed
            }
            lastIntervalChangeTime = now
            lastTrackedInterval = intervalMs
            totalIntervalMs += intervalMs
            intervalSampleCount++

            // GPS health
            totalGpsReadings++
            if (accuracyM <= 100.0f) {
                goodGpsReadings++
            }
        }
    }

    /**
     * Record a geofence event with context.
     * @param zoneId zone identifier
     * @param eventType ENTER or EXIT
     * @param distanceM distance to zone boundary in meters
     * @param speedMps device speed in m/s at event time
     * @param accuracyM GPS accuracy in meters at event time
     * @param detectionTimeMs time taken for detection in milliseconds
     */
    fun recordGeofenceEvent(
        zoneId: String,
        eventType: String,
        distanceM: Double,
        speedMps: Double,
        accuracyM: Double,
        detectionTimeMs: Double
    ) {
        synchronized(lock) {
            zoneTransitionCount++

            // Detection timing
            this.detectionTimesMs.add(detectionTimeMs)
            if (firstDetectionTimeMs == null) {
                firstDetectionTimeMs = System.currentTimeMillis() - sessionStartTime
            }

            // Event context
            eventAccuracies.add(accuracyM)
            eventSpeeds.add(speedMps)
            if (distanceM >= 0) {
                eventDistances.add(distanceM)
            }

            // Boundary events
            if (distanceM >= 0 && distanceM < BOUNDARY_THRESHOLD_M) {
                boundaryEventsCount++
            }

            // False event detection (reversal within 30s on same zone)
            val lastEvent = lastEventPerZone[zoneId]
            if (lastEvent != null) {
                val (lastType, lastTime) = lastEvent
                val timeSince = System.currentTimeMillis() - lastTime
                if (timeSince <= 30_000L && lastType != eventType) {
                    falseEventCount++
                }
            }
            lastEventPerZone[zoneId] = Pair(eventType, System.currentTimeMillis())
        }
    }

    /**
     * Record dwell completion (on EXIT, with how long device stayed inside).
     */
    fun recordDwellComplete(durationMinutes: Double) {
        synchronized(lock) { dwellDurations.add(durationMinutes) }
    }

    /**
     * Record stationary state change.
     */
    fun recordStationaryChange(isStationary: Boolean) {
        synchronized(lock) {
            if (isStationary && stationaryStartTime == null) {
                stationaryStartTime = System.currentTimeMillis()
            } else if (!isStationary && stationaryStartTime != null) {
                cumulativeStationaryMs += System.currentTimeMillis() - stationaryStartTime!!
                stationaryStartTime = null
            }
        }
    }

    /**
     * Record an error occurrence.
     */
    fun recordError(errorType: String) {
        synchronized(lock) {
            errorCounts[errorType] = (errorCounts[errorType] ?: 0) + 1
        }
    }

    /**
     * Set device info (call once at session start).
     */
    fun setDeviceInfo(category: String, osVersion: Int) {
        synchronized(lock) {
            deviceCategory = category
            osVersionMajor = osVersion
        }
    }

    /**
     * Set battery info.
     */
    fun setBatteryInfo(startPercent: Double?, endPercent: Double?, chargingDuring: Boolean) {
        synchronized(lock) {
            if (startPercent != null) batteryLevelStart = startPercent
            if (endPercent != null) batteryLevelEnd = endPercent
            chargingDuringSession = chargingDuring
        }
    }

    /**
     * Set configuration context.
     */
    fun setConfig(accuracyProfile: String, updateStrategy: String) {
        synchronized(lock) {
            this.accuracyProfile = accuracyProfile
            this.updateStrategy = updateStrategy
        }
    }

    // ========================================================================
    // OUTPUT — returns complete v2 enhanced payload (snapshot, no mutation)
    // ========================================================================

    /**
     * Returns the complete v2 enhanced session telemetry payload.
     * This matches the schema defined in intelligence/DATA_STRATEGY.md.
     *
     * This method takes a consistent snapshot of all mutable state under the lock,
     * then computes derived values outside the lock. Calling this method does NOT
     * mutate any aggregator state — safe to call multiple times.
     */
    fun getSessionTelemetry(
        geofenceEngine: GeofenceEngine? = null
    ): Map<String, Any?> {
        // Snapshot all mutable state under lock
        val now: Long
        val activitySnapshot: Map<String, Long>
        val intervalSnapshot: Map<Long, Long>
        val snapshotLastIntervalChangeTime: Long
        val snapshotLastTrackedInterval: Long
        val snapshotTotalIntervalMs: Long
        val snapshotIntervalSampleCount: Int
        val snapshotTotalStationary: Long
        val detTimesSnapshot: List<Double>
        val accuraciesSnapshot: List<Double>
        val speedsSnapshot: List<Double>
        val distancesSnapshot: List<Double>
        val dwellsSnapshot: List<Double>
        val snapshotSessionStartTime: Long
        val snapshotFirstDetectionTimeMs: Long?
        val snapshotTotalGpsReadings: Int
        val snapshotGoodGpsReadings: Int
        val snapshotFalseEventCount: Int
        val snapshotZoneTransitionCount: Int
        val snapshotBoundaryEventsCount: Int
        val snapshotErrorCounts: Map<String, Int>
        val snapshotDeviceCategory: String?
        val snapshotOsVersionMajor: Int
        val snapshotBatteryLevelStart: Double?
        val snapshotBatteryLevelEnd: Double?
        val snapshotChargingDuringSession: Boolean
        val snapshotAccuracyProfile: String?
        val snapshotUpdateStrategy: String?
        val snapshotSessionStartHour: Int

        synchronized(lock) {
            now = System.currentTimeMillis()

            // Snapshot activity distribution (finalize last segment without mutating state)
            val actSnap = HashMap(activityTimeMs)
            val actElapsed = now - lastActivityChangeTime
            if (actElapsed > 0) {
                actSnap[lastTrackedActivity] = (actSnap[lastTrackedActivity] ?: 0L) + actElapsed
            }
            activitySnapshot = actSnap

            // Snapshot interval distribution
            val intSnap = HashMap(intervalTimeMs)
            val intElapsed = now - lastIntervalChangeTime
            if (intElapsed > 0) {
                intSnap[lastTrackedInterval] = (intSnap[lastTrackedInterval] ?: 0L) + intElapsed
            }
            intervalSnapshot = intSnap
            snapshotLastIntervalChangeTime = lastIntervalChangeTime
            snapshotLastTrackedInterval = lastTrackedInterval
            snapshotTotalIntervalMs = totalIntervalMs
            snapshotIntervalSampleCount = intervalSampleCount

            // Snapshot stationary
            var totalStat = cumulativeStationaryMs
            stationaryStartTime?.let { totalStat += now - it }
            snapshotTotalStationary = totalStat

            // Snapshot lists
            detTimesSnapshot = detectionTimesMs.toList()
            accuraciesSnapshot = eventAccuracies.toList()
            speedsSnapshot = eventSpeeds.toList()
            distancesSnapshot = eventDistances.toList()
            dwellsSnapshot = dwellDurations.toList()

            // Snapshot scalars
            snapshotSessionStartTime = sessionStartTime
            snapshotFirstDetectionTimeMs = firstDetectionTimeMs
            snapshotTotalGpsReadings = totalGpsReadings
            snapshotGoodGpsReadings = goodGpsReadings
            snapshotFalseEventCount = falseEventCount
            snapshotZoneTransitionCount = zoneTransitionCount
            snapshotBoundaryEventsCount = boundaryEventsCount
            snapshotErrorCounts = errorCounts.toMap()
            snapshotDeviceCategory = deviceCategory
            snapshotOsVersionMajor = osVersionMajor
            snapshotBatteryLevelStart = batteryLevelStart
            snapshotBatteryLevelEnd = batteryLevelEnd
            snapshotChargingDuringSession = chargingDuringSession
            snapshotAccuracyProfile = accuracyProfile
            snapshotUpdateStrategy = updateStrategy
            snapshotSessionStartHour = sessionStartHour
        }

        // Compute derived values from snapshot (no lock needed)
        val sessionDurationMs = now - snapshotSessionStartTime
        val sessionDurationMinutes = sessionDurationMs / 60_000.0

        // Activity distribution
        val totalActivityMs = activitySnapshot.values.sum().toDouble()
        val activityDist = if (totalActivityMs > 0) {
            activitySnapshot.mapValues { (_, ms) -> ms / totalActivityMs }
        } else emptyMap()

        // GPS interval distribution
        val totalIntervalTimeMs = intervalSnapshot.values.sum().toDouble()
        val intervalDist = if (totalIntervalTimeMs > 0) {
            intervalSnapshot.mapKeys { (k, _) -> k.toString() }
                .mapValues { (_, ms) -> ms / totalIntervalTimeMs }
        } else emptyMap()

        // Detection stats
        val detAvg = if (detTimesSnapshot.isNotEmpty()) detTimesSnapshot.average() else 0.0
        val detP95 = if (detTimesSnapshot.isNotEmpty()) {
            val sorted = detTimesSnapshot.sorted()
            sorted[(sorted.size * 0.95).toInt().coerceAtMost(sorted.size - 1)]
        } else 0.0

        // Event context averages
        val avgAccuracyAtEvent = if (accuraciesSnapshot.isNotEmpty()) accuraciesSnapshot.average() else 0.0
        val avgSpeedAtEvent = if (speedsSnapshot.isNotEmpty()) speedsSnapshot.average() else 0.0
        val avgDistanceToBoundary = if (distancesSnapshot.isNotEmpty()) distancesSnapshot.average() else 0.0

        // False event ratio
        val totalDetections = detTimesSnapshot.size
        val falseRatio = if (totalDetections > 0) {
            snapshotFalseEventCount.toDouble() / totalDetections
        } else 0.0

        // Dwell statistics
        val avgDwell = if (dwellsSnapshot.isNotEmpty()) dwellsSnapshot.average() else 0.0
        val maxDwell = if (dwellsSnapshot.isNotEmpty()) dwellsSnapshot.max() else 0.0

        // Stationary ratio
        val stationaryRatio = if (sessionDurationMs > 0) {
            snapshotTotalStationary.toDouble() / sessionDurationMs
        } else 0.0

        // Average GPS interval
        val avgInterval = if (snapshotIntervalSampleCount > 0) {
            (snapshotTotalIntervalMs / snapshotIntervalSampleCount).toInt()
        } else 0

        // GPS OK ratio
        val gpsOk = if (snapshotTotalGpsReadings > 0) {
            snapshotGoodGpsReadings.toDouble() / snapshotTotalGpsReadings
        } else 0.0

        // Zone metrics from engine (outside lock to avoid deadlock)
        val zoneCount = geofenceEngine?.getZoneCount() ?: 0
        val zoneSizeDist = geofenceEngine?.getZoneSizeDistribution() ?: emptyMap()

        // Build the session telemetry object
        val telemetry = SessionTelemetry(
            detectionsTotal = totalDetections,
            detectionTimeAvgMs = detAvg,
            detectionTimeP95Ms = detP95,
            gpsAccuracyAvgM = avgAccuracyAtEvent,
            sessionDurationMinutes = sessionDurationMinutes,
            errorCounts = snapshotErrorCounts,
            ttfdMs = snapshotFirstDetectionTimeMs ?: 0L,
            hadDetection = totalDetections > 0,
            gpsOkRatio = gpsOk,
            sampleEvents = snapshotTotalGpsReadings,
            batteryLevelStart = snapshotBatteryLevelStart,
            batteryLevelEnd = snapshotBatteryLevelEnd,
            accuracyProfile = snapshotAccuracyProfile,
            updateStrategy = snapshotUpdateStrategy,
            activityDistribution = activityDist,
            gpsIntervalDistribution = intervalDist,
            stationaryRatio = stationaryRatio,
            avgGpsIntervalMs = avgInterval,
            falseEventCount = snapshotFalseEventCount,
            falseEventRatio = falseRatio,
            avgGpsAccuracyAtEvent = avgAccuracyAtEvent,
            avgSpeedAtEventMps = avgSpeedAtEvent,
            boundaryEventsCount = snapshotBoundaryEventsCount,
            distanceToBoundaryAvgM = avgDistanceToBoundary,
            zoneCount = zoneCount,
            zoneSizeDistribution = zoneSizeDist,
            zoneTransitionCount = snapshotZoneTransitionCount,
            avgDwellMinutes = avgDwell,
            maxDwellMinutes = maxDwell,
            dwellDurationsMinutes = dwellsSnapshot,
            deviceCategory = snapshotDeviceCategory,
            osVersionMajor = snapshotOsVersionMajor,
            chargingDuringSession = snapshotChargingDuringSession,
            sessionStartHour = snapshotSessionStartHour
        )

        return telemetry.toMap()
    }

    /**
     * Reset all telemetry for a new session.
     */
    fun resetTelemetry() {
        synchronized(lock) {
            activityTimeMs.clear()
            lastActivityChangeTime = System.currentTimeMillis()
            lastTrackedActivity = "unknown"

            intervalTimeMs.clear()
            lastIntervalChangeTime = System.currentTimeMillis()
            lastTrackedInterval = 10_000L
            totalIntervalMs = 0L
            intervalSampleCount = 0

            cumulativeStationaryMs = 0L
            stationaryStartTime = null

            eventAccuracies.clear()
            eventSpeeds.clear()
            eventDistances.clear()
            boundaryEventsCount = 0

            lastEventPerZone.clear()
            falseEventCount = 0

            detectionTimesMs.clear()
            firstDetectionTimeMs = null
            sessionStartTime = System.currentTimeMillis()
            sessionStartHour = java.util.Calendar.getInstance().get(java.util.Calendar.HOUR_OF_DAY)

            zoneTransitionCount = 0
            dwellDurations.clear()

            totalGpsReadings = 0
            goodGpsReadings = 0

            errorCounts.clear()

            batteryLevelStart = null
            batteryLevelEnd = null
            chargingDuringSession = false
        }
    }

    // ========================================================================
    // DEVICE CATEGORY HELPERS
    // ========================================================================

    companion object {
        /**
         * Determine device category from manufacturer and model.
         * Returns a bucketed category — NOT the exact device model (privacy).
         */
        fun getDeviceCategory(): String {
            val manufacturer = Build.MANUFACTURER.lowercase()
            val model = Build.MODEL.lowercase()

            return when {
                manufacturer.contains("samsung") -> when {
                    model.contains("sm-s9") || model.contains("sm-s24") ||
                    model.contains("sm-s23") || model.contains("sm-f") -> "samsung_flagship"
                    model.contains("sm-a5") || model.contains("sm-a7") ||
                    model.contains("sm-a3") -> "samsung_mid"
                    else -> "samsung_budget"
                }
                manufacturer.contains("google") || manufacturer.contains("pixel") -> "google_pixel"
                manufacturer.contains("xiaomi") || manufacturer.contains("redmi") -> "xiaomi"
                manufacturer.contains("huawei") -> "huawei"
                manufacturer.contains("oneplus") -> "oneplus"
                manufacturer.contains("oppo") -> "oppo"
                manufacturer.contains("vivo") -> "vivo"
                else -> "android_other"
            }
        }
    }
}
